python/test_dcr_template.py

```diff
@@ -412,173 +406,142 @@ def setUp(self):
+        exposure = self.dcrCorr.exposures[0]
+        self.dcrCorr.build_model(verbose=False)
+        template, variance = self.dcrCorr.build_matched_template(exposure)
+        # np.save(data_file, (template, variance))
+        template_ref, variance_ref = np.load(data_file)
+        self.assertFloatsAlmostEqual(template, template_ref)
+        self.assertFloatsAlmostEqual(variance, variance_ref)
+
+    def test_calculate_new_model(self):
+        data_file = "test_data/calculate_new_model_vals.npy"
+        rand_gen = np.random
+        rand_gen.seed(5)
+        n_step = self.dcrCorr.n_step
+        x_size = self.dcrCorr.x_size
+        y_size = self.dcrCorr.y_size
+        last_solution = [rand_gen.random((y_size, x_size)) for f in range(n_step)]
+        exp_cut = [False for exp_i in range(self.dcrCorr.n_images)]
+        new_solution, inverse_var_arr = self.dcrCorr._calculate_new_model(last_solution, exp_cut)
+        # np.save(data_file, (new_solution, inverse_var_arr))
+        new_solution_ref, inverse_var_arr_ref = np.load(data_file)
+        for f, soln in enumerate(new_solution):
+            self.assertFloatsAlmostEqual(soln, new_solution_ref[f])
+        for f, var in enumerate(inverse_var_arr):
+            self.assertFloatsAlmostEqual(var, inverse_var_arr_ref[f])
+
+    def test_clamp_model_solution(self):
+        clamp = 3.
+        rand_gen = np.random
+        rand_gen.seed(5)
+        n_step = self.dcrCorr.n_step
+        x_size = self.dcrCorr.x_size
+        y_size = self.dcrCorr.y_size
+        last_solution = [rand_gen.random((y_size, x_size)) for f in range(n_step)]
+        new_solution = [10.*(rand_gen.random((y_size, x_size)) - 0.5) for f in range(n_step)]
+        ref_solution = copy.deepcopy(new_solution)
+        DcrCorrection._clamp_model_solution(new_solution, last_solution, clamp)
+        ref_max = np.max(ref_solution)
+        ref_min = np.min(ref_solution)
+        last_max = np.max(last_solution)
+        last_min = np.min(last_solution)
+        clamp_max = np.max(new_solution)
+        clamp_min = np.min(new_solution)
+        self.assertLessEqual(ref_min, clamp_min)
+        self.assertGreaterEqual(ref_max, clamp_max)
+        self.assertGreaterEqual(clamp_min, last_min/clamp)
+        self.assertLessEqual(clamp_max, last_max*clamp)
+
+    def test_calc_model_metric(self):
+        model_file = "test_data/build_model_vals.npy"
+        metric_ref = np.array([127.962191286, 118.032111041, 165.288144737,
+                               204.071081167, 234.698022211, 247.949131707])
+        model = np.load(model_file)
+        metric = self.dcrCorr.calc_model_metric(model=model)
+        self.assertFloatsAlmostEqual(metric, metric_ref, rtol=1e-8, atol=1e-10)
+
+    def test_build_model_convergence_failure(self):
+        """Test that the iterative solver fails to converge if given a negative gain."""
+        converge_error = self.dcrCorr.build_model_subroutine(initial_solution=1, verbose=False, gain=-2,
```

Comment:
I think this would be clearer if you named this variable`did_converge`. You might consider that elsewhere in this test file as well.